### ISSUE TITLE
[flutter_tools] pin frontend_server_client because upgrading from 3.2.0 -> 4.0.0 regresses test

### ISSUE THREAD

**Issue by christopherfujino** [flutter/flutter#146164]
Example failing build: https://ci.chromium.org/p/flutter/builders/try/Windows%20tool_integration_tests_2_6/26698?

From attempted roll: https://github.com/flutter/flutter/pull/145959

Upstream tracking issue: https://github.com/dart-lang/webdev/issues/2377

**Comment by christopherfujino** [comment]
I ran this locally and added debug prints on the stdout and stderr of the `dart.bat` invocation:

```
[stderr] Checking Dart SDK version...

[stderr] Downloading Dart SDK from Flutter engine ...

[stderr] Rename-Item : Access to the path 'D:\git\flutter\bin\cache\dart-sdk' is denied.
At D:\git\flutter\bin\internal\update_dart_sdk.ps1:74 char:9
+         Rename-Item $dartSdkPath "$oldDartSdkPrefix$oldDartSdkSuffix"
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (D:\git\flutter\bin\cache\dart-sdk:String) [Rename-Item], IOException

[stderr]     + FullyQualifiedErrorId : RenameItemIOError,Microsoft.PowerShell.Commands.RenameItemCommand

[stderr]

[stderr] Error: Unable to update Dart SDK. Retrying...

[stderr] ERROR: Input redirection is not supported, exiting the process immediately.

[stderr] Checking Dart SDK version...

[stderr] Downloading Dart SDK from Flutter engine ...
```

**Comment by derekxu16** [comment]
The timeout was happening for me locally because `batch_entrypoint_test.dart` tries to rename `frontend_server_aot.dart.snapshot`, but this happens when `frontend_server_aot.dart.snapshot` is actively in use, because it is used by the test runner, which makes `Rename-Item` fail. The changes in https://github.com/flutter/flutter/pull/146229 make the test pass locally, but the CI check is still timing out, so there must be something else going wrong.

**Comment by derekxu16** [comment]
I added some logging statements to https://github.com/flutter/flutter/pull/146229 and observed that the test times out on CI because a file is not getting copied successfully. Excerpt from [these logs](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8751489754652583569/+/u/run_test.dart_for_tool_integration_tests_shard_and_subshard_2_6/stdout):

```
../../../../t/flutter_batch_entrypoint_test.d8ce6ed5/flutter/packages/flutter_tools/bin/flutter_tools.dart:5:8: Error: Error when reading '../../../../t/flutter_batch_entrypoint_test.d8ce6ed5/flutter/packages/flutter_tools/lib/executable.dart': The system cannot find the path specified.


import 'package:flutter_tools/executable.dart' as executable;
       ^
../../../../t/flutter_batch_entrypoint_test.d8ce6ed5/flutter/packages/flutter_tools/bin/flutter_tools.dart:8:14: Error: Method not found: 'main'.
  executable.main(args);
             ^^^^

Error: Unable to create dart snapshot for flutter tool.
```

I updated the PR with the following additional expectation to confirm this:

https://github.com/flutter/flutter/blob/2bd2ef93389c1ba649c43ac0328c049e096a2995/packages/flutter_tools/test/integration.shard/batch_entrypoint_test.dart#L68-L76

Excerpt from [these logs](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8751484197502462305/+/u/run_test.dart_for_tool_integration_tests_shard_and_subshard_2_6/stdout):

```
22:45 +11 ~2 -1: test/integration.shard/batch_entrypoint_test.dart: flutter/bin/dart updates the Dart SDK without hanging [E]
  Expected: <true>
    Actual: <false>
  
  package:matcher                                         expect
  test\integration.shard\batch_entrypoint_test.dart 68:5  main.<fn>
```

One solution is to make `batch_entrypoint_test.dart` do a clone of the flutter/flutter repo into the temp folder instead of trying to copy it there. Is that acceptable, @christopherfujino? Or do you have another idea?

**Comment by christopherfujino** [comment]
> I added some logging statements to #146229 and observed that the test times out on CI because a file is not getting copied successfully. Excerpt from [these logs](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8751489754652583569/+/u/run_test.dart_for_tool_integration_tests_shard_and_subshard_2_6/stdout):
> 
> ```
> ../../../../t/flutter_batch_entrypoint_test.d8ce6ed5/flutter/packages/flutter_tools/bin/flutter_tools.dart:5:8: Error: Error when reading '../../../../t/flutter_batch_entrypoint_test.d8ce6ed5/flutter/packages/flutter_tools/lib/executable.dart': The system cannot find the path specified.
> 
> 
> import 'package:flutter_tools/executable.dart' as executable;
>        ^
> ../../../../t/flutter_batch_entrypoint_test.d8ce6ed5/flutter/packages/flutter_tools/bin/flutter_tools.dart:8:14: Error: Method not found: 'main'.
>   executable.main(args);
>              ^^^^
> 
> Error: Unable to create dart snapshot for flutter tool.
> ```
> 
> I updated the PR with the following additional expectation to confirm this:
> 
> https://github.com/flutter/flutter/blob/2bd2ef93389c1ba649c43ac0328c049e096a2995/packages/flutter_tools/test/integration.shard/batch_entrypoint_test.dart#L68-L76
> 
> Excerpt from [these logs](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8751484197502462305/+/u/run_test.dart_for_tool_integration_tests_shard_and_subshard_2_6/stdout):
> 
> ```
> 22:45 +11 ~2 -1: test/integration.shard/batch_entrypoint_test.dart: flutter/bin/dart updates the Dart SDK without hanging [E]
>   Expected: <true>
>     Actual: <false>
>   
>   package:matcher                                         expect
>   test\integration.shard\batch_entrypoint_test.dart 68:5  main.<fn>
> ```
> 
> One solution is to make `batch_entrypoint_test.dart` do a clone of the flutter/flutter repo into the temp folder instead of trying to copy it there. Is that acceptable, @christopherfujino? Or do you have another idea?

I'm taking a look at this now...

**Comment by christopherfujino** [comment]
So after investigating this issue and the proposed fix in https://github.com/flutter/flutter/pull/146229, here are my conclusions:

1. The disk i/o implementations of `dart.exe foo.appjit.snapshot` and `dartaotruntime.exe foo.aot.snapshot` are different in that the latter will keep the snapshot file open for the duration of VM execution while the former will (presumably) close it after having read the data. I verified this with running a minimal dart program that runs indefinitely from both appjit and aot snapshots, and from another powershell session tried calling `Rename-Item` on the snapshot file--the rename succeeded on the appjit snapshot but failed on aot snapshot with `Rename-Item : The process cannot access the file because it is being used by another process.`.
2. version 4.0.0 of `frontend_server_client` is changing the package:test runner to use an aot snapshot of the frontend_server rather than an appjit snapshot: https://github.com/dart-lang/webdev/pull/2263.
3. The tool's [batch_entrypoint_test](https://github.com/flutter/flutter/blob/75ae44df77aef3d109add07abfd88438732bbee8/packages/flutter_tools/test/integration.shard/batch_entrypoint_test.dart) is validating the logic in Flutter's [update_dart_sdk.ps1](https://github.com/flutter/flutter/blob/75ae44df77aef3d109add07abfd88438732bbee8/bin/internal/update_dart_sdk.ps1) powershell script. Of interest to this issue is that at [line 74](https://github.com/flutter/flutter/blob/75ae44df77aef3d109add07abfd88438732bbee8/bin/internal/update_dart_sdk.ps1#L74), before downloading a new version of the Dart SDK, we call `Rename-Item` on the existing (or old) Dart SDK to move it to a path with `dart-sdk.old` in the name. I believe we do this because just deleting the old one while the tool is running would fail.
4. The fact that this test is failing after this change to the frontend_server_client depended on by the test runner is incidental to the specific behavior the test is running, and thus Derek's proposed test changes would make sense there.
5. The fact that this test is failing inadvertently revealed what I believe would be a real customer regression, in that when calling `flutter upgrade`, if there is any running dart process that is using an aot snapshot from the Dart SDK within the Flutter SDK's cache, `flutter upgrade` would likely fail at the step where it tries to download the new version of the Dart SDK. Re-running the flutter CLI tool would likely repeatedly fail as long as said Dart process is running.

It seems like the immediately implications of rolling `frontend_server_client` forward will be that--on windows--any user using `flutter upgrade` while a dart process spawned via `frontend_server_client` is running will experience a crash (of the flutter upgrade process), and likely not be able to download the new Dart SDK until the other dart process exits.

@elliette & @derekxu16 do you know if anything else in Flutter is using frontend_server_client other than testing?

**Comment by christopherfujino** [comment]
For context, I filed a more general tracking issue for how this will affect Flutter users: https://github.com/flutter/flutter/issues/146164#issuecomment-2040805169

**Comment by derekxu16** [comment]
> @elliette & @derekxu16 do you know if anything else in Flutter is using frontend_server_client other than testing?

I don't know of anything else in Flutter that uses `package:frontend_server_client`. This commit (https://github.com/flutter/flutter/commit/1a44710d19809134fac64ab3fc3cf112e48b0868) added some sites where `flutter_tools` launches processes that run the `frontend_server` AOT snapshot though. These processes never outlive the invocation of the `flutter` tool that starts them.

**Comment by derekxu16** [comment]
Fixed by https://github.com/flutter/flutter/pull/146650

**Comment by christopherfujino** [comment]
> Fixed by #146650

Thanks!
